File: android/app/src/main/java/com/turbometa/rayban/viewmodels/OmniRealtimeViewModel.kt
package com.turbometa.rayban.viewmodels

import android.app.Application
import android.graphics.Bitmap
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.turbometa.rayban.data.ConversationStorage
import com.turbometa.rayban.managers.APIProviderManager
import com.turbometa.rayban.models.ConversationMessage
import com.turbometa.rayban.models.ConversationRecord
import com.turbometa.rayban.models.MessageRole
import com.turbometa.rayban.services.GeminiLiveService
import com.turbometa.rayban.utils.APIKeyManager
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.UUID

/**
 * OmniRealtimeViewModel
 * Uses Google Gemini Live for real-time AI conversation
 * 1:1 port from iOS OmniRealtimeViewModel
 */
class OmniRealtimeViewModel(application: Application) : AndroidViewModel(application) {

    companion object {
        private const val TAG = "OmniRealtimeViewModel"
    }

    private val apiKeyManager = APIKeyManager.getInstance(application)
    private val providerManager = APIProviderManager.getInstance(application)
    private val conversationStorage = ConversationStorage.getInstance(application)

    // Service
    private var geminiService: GeminiLiveService? = null

    // State
    sealed class ViewState {
        object Idle : ViewState()
        object Connecting : ViewState()
        object Connected : ViewState()
        object Recording : ViewState()
        object Processing : ViewState()
        object Speaking : ViewState()
        data class Error(val message: String) : ViewState()
    }

    private val _viewState = MutableStateFlow<ViewState>(ViewState.Idle)
    val viewState: StateFlow<ViewState> = _viewState.asStateFlow()

    private val _messages = MutableStateFlow<List<ConversationMessage>>(emptyList())
    val messages: StateFlow<List<ConversationMessage>> = _messages.asStateFlow()

    private val _currentTranscript = MutableStateFlow("")
    val currentTranscript: StateFlow<String> = _currentTranscript.asStateFlow()

    private val _userTranscript = MutableStateFlow("")
    val userTranscript: StateFlow<String> = _userTranscript.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _isRecording = MutableStateFlow(false)
    val isRecording: StateFlow<Boolean> = _isRecording.asStateFlow()

    private val _isSpeaking = MutableStateFlow(false)
    val isSpeaking: StateFlow<Boolean> = _isSpeaking.asStateFlow()

    // Periodic image sending
    private val _imageSendInterval = MutableStateFlow(1.0)
    val imageSendInterval: StateFlow<Double> = _imageSendInterval.asStateFlow()

    private var imageSendJob: Job? = null
    private var isImageSendingEnabled = false
    private var currentVideoFrame: Bitmap? = null

    private var currentSessionId: String = UUID.randomUUID().toString()
    private var pendingVideoFrame: Bitmap? = null

    init {
        initializeService()
    }

    private fun initializeService() {
        val apiKey = providerManager.getLiveAIAPIKey(apiKeyManager)
        val language = apiKeyManager.getOutputLanguage()

        if (apiKey.isBlank()) {
            _errorMessage.value = "API Key not configured for Google Gemini"
            Log.e(TAG, "API Key not configured for Google Gemini")
            return
        }

        Log.d(TAG, "Initializing Gemini Live service")

        val model = APIProviderManager.liveAIDefaultModel

        geminiService = GeminiLiveService(apiKey, model, language, getApplication()).apply {
            onTranscriptDelta = { delta ->
                _currentTranscript.value += delta
            }

            onTranscriptDone = { transcript ->
                if (transcript.isNotBlank()) {
                    addAssistantMessage(_currentTranscript.value.ifBlank { transcript })
                }
                _currentTranscript.value = ""
                _viewState.value = ViewState.Connected
            }

            onUserTranscript = { transcript ->
                if (transcript.isNotBlank()) {
                    _userTranscript.value = transcript
                    addUserMessage(transcript)
                }
            }

            onSpeechStarted = {
                _viewState.value = ViewState.Recording
            }

            onSpeechStopped = {
                _viewState.value = ViewState.Processing
            }

            onError = { error ->
                _errorMessage.value = error
                _viewState.value = ViewState.Error(error)
            }

            onConnected = {
                _isConnected.value = true
                _viewState.value = ViewState.Connected
            }

            onFirstAudioSent = {
                // Start periodic image sending after a 1s delay (matching iOS)
                viewModelScope.launch {
                    Log.d(TAG, "Received first audio send callback, starting periodic image sending")
                    delay(1000)
                    isImageSendingEnabled = true
                    startImageSendTimer()
                    Log.d(TAG, "Periodic image sending started")
                }
            }
        }

        observeGeminiServiceStates()
    }

    private fun observeGeminiServiceStates() {
        viewModelScope.launch {
            geminiService?.isConnected?.collect { connected ->
                _isConnected.value = connected
                if (connected && _viewState.value == ViewState.Connecting) {
                    _viewState.value = ViewState.Connected
                } else if (!connected && _viewState.value != ViewState.Idle) {
                    _viewState.value = ViewState.Idle
                }
            }
        }

        viewModelScope.launch {
            geminiService?.isRecording?.collect { recording ->
                _isRecording.value = recording
            }
        }

        viewModelScope.launch {
            geminiService?.isSpeaking?.collect { speaking ->
                _isSpeaking.value = speaking
                if (speaking) {
                    _viewState.value = ViewState.Speaking
                }
            }
        }
    }

    // MARK: - Periodic Image Sending

    private fun startImageSendTimer() {
        stopImageSendTimer()
        imageSendJob = viewModelScope.launch {
            while (true) {
                delay(((_imageSendInterval.value) * 1000).toLong())
                if (isImageSendingEnabled) {
                    currentVideoFrame?.let { frame ->
                        geminiService?.sendImageInput(frame)
                    }
                }
            }
        }
    }

    private fun stopImageSendTimer() {
        imageSendJob?.cancel()
        imageSendJob = null
    }

    fun setImageSendInterval(interval: Double) {
        _imageSendInterval.value = interval
        // Restart timer with new interval if currently sending
        if (isImageSendingEnabled) {
            startImageSendTimer()
        }
    }

    fun connect() {
        viewModelScope.launch {
            if (_isConnected.value) return@launch

            _viewState.value = ViewState.Connecting
            _messages.value = emptyList()
            currentSessionId = UUID.randomUUID().toString()

            geminiService?.connect()
        }
    }

    fun disconnect() {
        viewModelScope.launch {
            saveCurrentConversation()
            stopImageSendTimer()
            isImageSendingEnabled = false
            geminiService?.disconnect()
            _viewState.value = ViewState.Idle
            _isConnected.value = false
            _messages.value = emptyList()
            _currentTranscript.value = ""
            _userTranscript.value = ""
        }
    }

    fun startRecording() {
        if (!_isConnected.value) {
            _errorMessage.value = "Not connected"
            return
        }

        // Update video frame if available
        pendingVideoFrame?.let { frame ->
            geminiService?.updateVideoFrame(frame)
        }

        geminiService?.startRecording()
        _viewState.value = ViewState.Recording
    }

    fun stopRecording() {
        geminiService?.stopRecording()
        stopImageSendTimer()
        isImageSendingEnabled = false
        if (_viewState.value == ViewState.Recording) {
            _viewState.value = ViewState.Processing
        }
    }

    fun updateVideoFrame(frame: Bitmap) {
        pendingVideoFrame = frame
        currentVideoFrame = frame
        geminiService?.updateVideoFrame(frame)
    }

    fun sendImage(image: Bitmap) {
        geminiService?.sendImageInput(image)
    }

    private fun addUserMessage(text: String) {
        val message = ConversationMessage(
            id = UUID.randomUUID().toString(),
            role = MessageRole.USER,
            content = text,
            timestamp = System.currentTimeMillis()
        )
        _messages.value = _messages.value + message
    }

    private fun addAssistantMessage(text: String) {
        val message = ConversationMessage(
            id = UUID.randomUUID().toString(),
            role = MessageRole.ASSISTANT,
            content = text,
            timestamp = System.currentTimeMillis()
        )
        _messages.value = _messages.value + message
    }

    private fun saveCurrentConversation() {
        if (_messages.value.isEmpty()) return

        val record = ConversationRecord(
            id = currentSessionId,
            timestamp = System.currentTimeMillis(),
            messages = _messages.value,
            aiModel = APIProviderManager.liveAIDefaultModel,
            language = apiKeyManager.getOutputLanguage()
        )

        conversationStorage.saveConversation(record)
    }

    fun clearError() {
        _errorMessage.value = null
        geminiService?.clearError()
        if (_viewState.value is ViewState.Error) {
            _viewState.value = if (_isConnected.value) ViewState.Connected else ViewState.Idle
        }
    }

    fun refreshService() {
        disconnect()
        geminiService = null
        initializeService()
    }

    override fun onCleared() {
        super.onCleared()
        saveCurrentConversation()
        stopImageSendTimer()
        geminiService?.disconnect()
    }
}

File: CameraAccess/ViewModels/OmniRealtimeViewModel.swift
/*
 * Omni Realtime ViewModel
 * Manages real-time multimodal conversation with AI
 * Uses Google Gemini Live for real-time audio+video chat
 */

import Foundation
import SwiftUI
import AVFoundation

@MainActor
class OmniRealtimeViewModel: ObservableObject {

    // Published state
    @Published var isConnected = false
    @Published var isRecording = false
    @Published var isSpeaking = false
    @Published var currentTranscript = ""
    @Published var conversationHistory: [ConversationMessage] = []
    @Published var errorMessage: String?
    @Published var showError = false

    // Service
    private var geminiService: GeminiLiveService?
    private let apiKey: String

    // Video frame
    private var currentVideoFrame: UIImage?
    private var isImageSendingEnabled = false // Whether image sending is enabled (after first audio)
    private var imageSendTimer: Timer?
    @Published var imageSendInterval: TimeInterval = 1.0
    private var lastUserTranscript = ""
    private var lastAssistantTranscript = ""

    init(apiKey: String) {
        self.apiKey = apiKey
        self.geminiService = GeminiLiveService(apiKey: apiKey)
        setupCallbacks()
    }

    // MARK: - Setup

    private func setupCallbacks() {
        guard let geminiService = geminiService else { return }

        geminiService.onConnected = { [weak self] in
            Task { @MainActor in
                self?.isConnected = true
            }
        }

        geminiService.onFirstAudioSent = { [weak self] in
            Task { @MainActor in
                print("✅ [LiveAI-VM] Received first audio send callback, starting periodic image sending")
                DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
                    self?.isImageSendingEnabled = true
                    self?.startImageSendTimer()
                    print("📸 [LiveAI-VM] Periodic image sending started")
                }
            }
        }

        geminiService.onTranscriptDelta = { [weak self] (delta: String) in
            Task { @MainActor in
                guard let self else { return }
                let cleaned = delta.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !cleaned.isEmpty else { return }
                print("📝 [LiveAI-VM] AI speech fragment: \(cleaned)")
                if self.currentTranscript.isEmpty {
                    self.currentTranscript = cleaned
                } else if cleaned.hasPrefix(self.currentTranscript) {
                    // Some providers send a full running transcript on each delta.
                    self.currentTranscript = cleaned
                } else {
                    // Others send only incremental chunks.
                    self.currentTranscript += cleaned
                }
            }
        }

        geminiService.onUserTranscript = { [weak self] (userText: String) in
            Task { @MainActor in
                guard let self = self else { return }
                let cleaned = userText.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !cleaned.isEmpty else { return }
                guard cleaned != self.lastUserTranscript else { return }
                self.lastUserTranscript = cleaned
                print("💬 [LiveAI-VM] Saving user speech: \(cleaned)")
                self.conversationHistory.append(
                    ConversationMessage(role: .user, content: cleaned)
                )
            }
        }

        geminiService.onTranscriptDone = { [weak self] (fullText: String) in
            Task { @MainActor in
                guard let self = self else { return }
                let textToSave = fullText.isEmpty ? self.currentTranscript : fullText
                let cleaned = textToSave.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !cleaned.isEmpty else {
                    print("⚠️ [LiveAI-VM] AI response is empty, skipping save")
                    return
                }
                guard cleaned != self.lastAssistantTranscript else {
                    self.currentTranscript = ""
                    return
                }
                self.lastAssistantTranscript = cleaned
                print("💬 [LiveAI-VM] Saving AI response: \(cleaned)")
                self.conversationHistory.append(
                    ConversationMessage(role: .assistant, content: cleaned)
                )
                self.currentTranscript = ""
            }
        }

        geminiService.onAudioDone = { [weak self] in
            Task { @MainActor in
                self?.isSpeaking = false
            }
        }

        geminiService.onError = { [weak self] (error: String) in
            Task { @MainActor in
                self?.errorMessage = error
                self?.showError = true
            }
        }
    }

    // MARK: - Connection

    func connect() {
        geminiService?.connect()
    }

    func disconnect() {
        // Save conversation before disconnecting
        saveConversation()

        stopImageSendTimer()
        stopRecording()
        geminiService?.disconnect()

        isConnected = false
        isImageSendingEnabled = false
    }

    private func saveConversation() {
        // Only save if there's meaningful conversation
        guard !conversationHistory.isEmpty else {
            print("💬 [LiveAI] No conversation content, skipping save")
            return
        }

        let record = ConversationRecord(
            messages: conversationHistory,
            aiModel: APIProviderManager.liveAIDefaultModel,
            language: "en-US"
        )

        ConversationStorage.shared.saveConversation(record)
        print("💾 [LiveAI] Conversation saved: \(conversationHistory.count) messages")
    }

    // MARK: - Recording

    func startRecording() {
        guard isConnected else {
            print("⚠️ [LiveAI] Not connected, cannot start recording")
            errorMessage = "Please connect to server first"
            showError = true
            return
        }

        print("🎤 [LiveAI] Start recording (voice-triggered mode) - Provider: Google Gemini")
        geminiService?.startRecording()
        isRecording = true
    }

    func stopRecording() {
        print("🛑 [LiveAI] Stop recording")
        stopImageSendTimer()
        geminiService?.stopRecording()
        isRecording = false
    }

    // MARK: - Video Frames

    func updateVideoFrame(_ frame: UIImage) {
        currentVideoFrame = frame
    }

    func setImageSendInterval(_ interval: TimeInterval) {
        imageSendInterval = interval
        if isImageSendingEnabled {
            startImageSendTimer()
        }
    }

    private func startImageSendTimer() {
        stopImageSendTimer()
        imageSendTimer = Timer.scheduledTimer(withTimeInterval: imageSendInterval, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self = self,
                      self.isImageSendingEnabled,
                      let frame = self.currentVideoFrame else { return }
                self.geminiService?.sendImageInput(frame)
            }
        }
    }

    private func stopImageSendTimer() {
        imageSendTimer?.invalidate()
        imageSendTimer = nil
    }

    // MARK: - Cleanup

    func dismissError() {
        showError = false
    }

    nonisolated deinit {
        Task { @MainActor [weak geminiService] in
            geminiService?.disconnect()
        }
    }
}

// MARK: - Conversation Message

struct ConversationMessage: Identifiable {
    let id = UUID()
    let role: MessageRole
    let content: String
    let timestamp = Date()

    enum MessageRole {
        case user
        case assistant
    }
}
